drivers/dma/dma_silabs_ldma.c

```diff
@@ -0,0 +1,524 @@
+	uint32_t complete_callback_en;
+	atomic_t busy;
+	void *user_data;
+	dma_callback_t cb;
+	LDMA_TransferCfg_t xfer_config;
+	LDMA_Descriptor_t *desc;
+};
+
+struct dma_silabs_config {
+	void (*config_irq)(const struct device *dev);
+	const struct device *clock_dev;
+};
+
+struct dma_silabs_data {
+	struct dma_context dma_ctx;
+	struct dma_silabs_channel *dma_chan_table;
+	struct sys_mem_blocks *dma_desc_pool;
+};
+
+static int dma_silabs_get_blocksize(uint32_t src_blen, uint32_t dst_blen, uint32_t src_dsize)
+{
+	const static struct {
+		int native;
+		int efr;
+	} ldma_blocksize_map[] = {
+		{0x0001, ldmaCtrlBlockSizeUnit1},
+		{0x0002, ldmaCtrlBlockSizeUnit2},
+		{0x0003, ldmaCtrlBlockSizeUnit3},
+		{0x0004, ldmaCtrlBlockSizeUnit4},
+		{0x0006, ldmaCtrlBlockSizeUnit6},
+		{0x0008, ldmaCtrlBlockSizeUnit8},
+		{0x0010, ldmaCtrlBlockSizeUnit16},
+		{0x0020, ldmaCtrlBlockSizeUnit32},
+		{0x0040, ldmaCtrlBlockSizeUnit64},
+		{0x0080, ldmaCtrlBlockSizeUnit128},
+		{0x0100, ldmaCtrlBlockSizeUnit256},
+		{0x0200, ldmaCtrlBlockSizeUnit512},
+		{0x0400, ldmaCtrlBlockSizeUnit1024}
+	};
+	uint32_t arb_unit;
+
+	if (src_blen != dst_blen) {
+		LOG_ERR("Source burst length (%u) and destination burst length(%u) must be equal",
+			src_blen, dst_blen);
+		return -ENOTSUP;
+	}
+
+	if (src_blen % src_dsize) {
+		LOG_ERR("burst length (%u) and data size (%u) mismatch", src_blen, dst_blen);
+		return -EINVAL;
+	}
+
+	arb_unit = src_blen / src_dsize;
+
+	for (int i = 0; i < ARRAY_SIZE(ldma_blocksize_map); i++) {
+		if (ldma_blocksize_map[i].native == arb_unit) {
+			return ldma_blocksize_map[i].efr;
+		}
+	}
+	return -EINVAL;
+}
+
+static int dma_silabs_block_to_descriptor(struct dma_config *config,
+					  struct dma_silabs_channel *chan_conf,
+					  struct dma_block_config *block, LDMA_Descriptor_t *desc)
+{
+	int ret, src_size;
+
+	if (block->dest_scatter_count || block->source_gather_count ||
+	    block->source_gather_interval || block->dest_scatter_interval ||
+	    block->dest_reload_en || block->source_reload_en) {
+		return -ENOTSUP;
+	}
+
+	if ((block->source_gather_en || block->dest_scatter_en) && config->block_count == 1) {
+		LOG_WRN("DMA scatter_gather enabled but there is only one descriptor "
+			"configured");
+	}
+
+	memset(desc, 0, sizeof(*desc));
+
+	desc->xfer.structReq = 1;
+
+	if (config->source_data_size != config->dest_data_size) {
+		LOG_ERR("Source data size(%u) and destination data size(%u) must be equal",
+			config->source_data_size, config->dest_data_size);
+		return -ENOTSUP;
+	}
+
+	src_size = config->source_data_size;
+	desc->xfer.size = LOG2(src_size);
+	if (desc->xfer.size < ldmaCtrlSizeByte || desc->xfer.size > ldmaCtrlSizeWord) {
+		return -ENOTSUP;
+	}
+
+	if (block->block_size % config->source_data_size) {
+		desc->xfer.xferCnt = block->block_size / config->source_data_size;
+	} else {
+		desc->xfer.xferCnt = block->block_size / config->source_data_size - 1;
+	}
+
+	if (desc->xfer.xferCnt > LDMA_DESCRIPTOR_MAX_XFER_SIZE) {
```

Comment:
Same logic as above; the `xferCnt` member is a bitfield, so it's already truncated by the time the check is done.